### PR TITLE
#161647983 Fix signup error response

### PR DESCRIPTION
**Pull request by augustineezinwa**
#### What does this PR do?
- This PR fixes a bug on the signup error response object. The purpose of this PR is that the error response returned when a user input fails validations is more descriptive.
#### Description of Task to be completed?
- update tests
- restructure signup error response
- abstract validator methods
- introduce reusability on validator methods
#### How should this be manually tested?
- `npm test`
#### Any background context you want to provide?
- None
#### What are the relevant pivotal tracker stories?
- [#161647983](https://www.pivotaltracker.com/story/show/161647983)
#### Screenshots (if appropriate)
screen shot 2018-11-01 at 8 59 45 pm

#### Questions:
- None

### CODE REVIEW

File: server/middlewares/UserValidation.js [hunk 1]
```diff
@@ -16,20 +17,65 @@ class UserValidation {
...
     */
   static validateUserSignUp(req, res, next) {
-    let { fullName } = req.body;
-    const { avatarUrl, bio } = req.body;
-    req.checkBody('fullName', 'full name is too long').isLength({ max: 40 });
+    UserValidation.validateFullName(req);
+    UserValidation.validateEmail(req);
+    UserValidation.validatePassword(req);
+    UserValidation.validateAvatarUrl(req);
+    UserValidation.validateBio(req);
```
Comment (Darthrighteous) [inline server/middlewares/UserValidation.js]: This class looks much better now, good job on that.
Since we are not requiring the user to provide avatarUrl and bio at sign up, why did you add the validations here?

File: server/middlewares/UserValidation.js [hunk 2]
```diff
@@ -16,20 +17,65 @@ class UserValidation {
...
+    */
+  static validateEmail(req) {
     req.checkBody('email', 'please enter email').exists();
     req.checkBody('email', 'please enter a valid email').isEmail();
+  }
+
+  /**
+    * @description - This method validates the password
+    * @param {object} req - The request object
+    * @param {object} password - The request object
```
Comment (Darthrighteous) [inline server/middlewares/UserValidation.js]: this `password` argument doesn't exist on this function, I don't think you meant to put it here.

File: server/middlewares/UserValidation.js [hunk 3]
```diff
@@ -72,7 +171,7 @@ class UserValidation {
         if (foundEmail) {
           return res.status(409).json({
             errors: {
-              message: ['email is already in use']
+              email: ['email is already in use']
             }
```
Comment (Darthrighteous) [inline server/middlewares/UserValidation.js]: I don't think we should be returning an error inside this function, it makes it only useful for the signup route (or at least a route that requires the email to not be in use). 
The log in route, for example, can't use this function since we want the email to exist in our db to log in.

I suggest we return true or false, then custom functionality/error can be written from where it's being called.

File: server/middlewares/UserValidation.js (hunk 1 above)
Comment (augustineezinwa) [reply server/middlewares/UserValidation.js]: We do not want to expose any vulnerability of any kind on the signup route. If for example, someone finds out that a particular field on the user table was not validated for, that might be our system's Achilles heel. The `validateAvatarUrl()` and `validateBio()` only comes to action if a someone enters some characters on those fields. 
Therefore, I find it important that this validations remain in place.

File: server/middlewares/UserValidation.js (hunk 2 above)
Comment (augustineezinwa) [reply server/middlewares/UserValidation.js]: Thanks for the feedback. I have taken note.

File: server/middlewares/UserValidation.js (hunk 1 above)
Comment (augustineezinwa) [reply server/middlewares/UserValidation.js]: Ok. fine

Comment (sulenchy) [reply server/middlewares/UserValidation.js]: I suggest you only validate what is required at signup. If you check the database schema at https://app.quickdatabasediagrams.com/#/d/OnZQOZ, the user table includes other fields that are not required like googleId, facebookId etc Are you saying you will still need to include them in `validateSignUp` method.

Comment (Darthrighteous) [reply server/middlewares/UserValidation.js]: Alternatively, what we could do is check that the object keys passed in req.body are fullName, email, password. If anything else is passed, the request is bounced.

Comment (augustineezinwa) [reply server/middlewares/UserValidation.js]: Thanks, teammates for the feedback. I will implement your suggestions.
